Sources/HummingbirdCore/Server/HTTPServer.swift

```diff
@@ -22,162 +23,223 @@ import NIOTransportServices
         /// waiting on the server while it is not running will throw this
         case serverNotRunning
         /// the current connection is closing
         case connectionClosing
+        /// the server is shutting down
+        case serverShuttingDown
+        /// the server has shutdown
+        case serverShutdown
     }
 
     /// Initialize HTTP server
     /// - Parameters:
     ///   - group: EventLoopGroup server uses
     ///   - configuration: Configuration for server
-    public init(group: EventLoopGroup, configuration: Configuration) {
+    public init(
+        group: EventLoopGroup,
+        configuration: Configuration,
+        responder: HBHTTPResponder,
+        childChannelInitializer: HBChannelInitializer = HTTP1Channel(),
+        additionalChannelHandlers: @autoclosure @escaping @Sendable () -> [any RemovableChannelHandler] = [],
+        onServerRunning: @escaping @Sendable () async -> Void = {},
+        logger: Logger
+    ) {
         self.eventLoopGroup = group
         self.configuration = configuration
-        self.quiesce = nil
-        self.childChannelHandlers = .init()
-        // defaults to HTTP1
-        self.httpChannelInitializer = HTTP1ChannelInitializer()
-        // add idle read, write handlers
-        if let idleTimeoutConfiguration = configuration.idleTimeoutConfiguration {
-            self.childChannelHandlers.addHandler(
-                IdleStateHandler(
-                    readTimeout: idleTimeoutConfiguration.readTimeout,
-                    writeTimeout: idleTimeoutConfiguration.writeTimeout
-                )
+        self.state = .initial(
+            responder: responder,
+            childChannelInitializer: childChannelInitializer,
+            onServerRunning: onServerRunning
+        )
+        self.additionalChannelHandlers = additionalChannelHandlers
+        self.logger = logger
+    }
+
+    /// Start server
+    /// - Parameter responder: Object that provides responses to requests sent to the server
+    /// - Returns: EventLoopFuture that is fulfilled when server has started
+    public func start() async throws {
+        switch self.state {
+        case .initial(let responder, let childChannelInitializer, let onServerRunning):
+            self.state = .starting
+            let (channel, quiescingHelper) = try await self.makeServer(
+                httpChannelInitializer: childChannelInitializer,
+                responder: responder
             )
+            // check state again
+            switch self.state {
+            case .initial, .running:
+                preconditionFailure("We should only be running once")
+
+            case .starting:
+                self.state = .running(channel: channel, quiescingHelper: quiescingHelper)
+                await onServerRunning()
+
+            case .shuttingDown, .shutdown:
+                try await channel.close()
+            }
+
+        case .starting, .running:
+            fatalError("Unexpected state")
```

Comment:
While starting the same server twice is certainly not the expected scenario, I don't think crashing is the right call.